File: Go/recursed/colRecur.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
)

/* --------------------------------------------------------------------------------------------
* Struct colNum that contains 2 elements: number and sequence length
* --------------------------------------------------------------------------------------------*/
type colNum struct {
	num int64
	seqLength int64
}

/* --------------------------------------------------------------------------------------------
* Main()
* Takes command line arguments and pares it as an unsigned Int64
* Fills up array and prints
* --------------------------------------------------------------------------------------------*/
func main() {
	var colArr[10] colNum
	var colArrSize int = 0

	lower, _ := strconv.Atoi(os.Args[1])
	upper, _ := strconv.Atoi(os.Args[2])

	for i := lower; i < upper; i++ {
		colValue := colNum{num: int64(i), seqLength: 0}
		colValue.seqLength = collatz(colValue.num, colValue.seqLength)
		if !replaceDuplicate(colArr, colValue, colArrSize) {
			if colArrSize < 10 {
				colArr[colArrSize] = colValue
				colArrSize++;
			} else {
				var location int  = searchMin(colArr, colArrSize)
				if colArr[location].seqLength < colValue.seqLength {
					colArr[location] = colValue
				}
			}
		}
	}

	printNums(colArr, colArrSize, 0)
	fmt.Println()
	printNums(colArr, colArrSize, 1)

}

/* --------------------------------------------------------------------------------------------
* Calculates the amount of numbers required to reach 1 in a collatz sequence
* Until num is 1, check if even or off and return collatz accordingly
* Utilizes recursion
* Returns the count
* --------------------------------------------------------------------------------------------*/
func collatz(num int64, count int64) int64 {

	if (num == 1) { return count }
	count++
	if (num%2 == 0) { return collatz(num/2, count) }
	return collatz((num*3)+1, count)
}

/* --------------------------------------------------------------------------------------------
* Replaces any duplicates in the array according to if num is less than or equal to element
* Returns true or false depending on if a duplicate was found
* --------------------------------------------------------------------------------------------*/
func replaceDuplicate(colArr[10] colNum, colValue colNum, arrSize int) bool {
	for i := 0; i < arrSize; i++ {
		if colArr[i].seqLength == colValue.seqLength {
			if colValue.num <= colArr[i].num { colArr[i] = colValue }
			return true
		}
	}
	return false
}

/* --------------------------------------------------------------------------------------------
* Sorts the array with Insertion Sort
* Based on the sortType passed through, it sorts according to num or seqLength
* Returns the sorted array
* --------------------------------------------------------------------------------------------*/
func searchMin(colArr[10] colNum, arrSize int) int {

	var minSt colNum = colArr[0]
	var minNorm float64 = 0
	var location int = 0
	
	for i := 0; i < arrSize; i++ {
	  minNorm = math.Min(float64(minSt.seqLength), float64(colArr[i].seqLength))
	  
	  if int64(minNorm) == colArr[i].seqLength {
		minSt = colArr[i]
		location = i
	  }
	}
	return location
}
  
/* --------------------------------------------------------------------------------------------
* Sorts the array with Insertion Sort
* Based on the sortType passed through, it sorts according to num or seqLength
* Returns the sorted array
* --------------------------------------------------------------------------------------------*/
func sort(colArr[10] colNum, arrSize int, sortType int) [10]colNum {

if sortType == 0 {
	fmt.Println("Sorted based on sequence length")
	var maxim colNum

	for i := 0; i < arrSize; i++ {
	var indx int = i
	
	for j := i; j < arrSize; j++ {
		if colArr[j].seqLength > colArr[indx].seqLength {
		indx = j
		}
	}
	maxim = colArr[indx]
	colArr[indx] = colArr[i]
	colArr[i] = maxim
	}
	return colArr
} else {
	fmt.Println("Sorted based on sequence size")
	var maxim colNum

	for i := 0; i < arrSize; i++ {
	var indx int = i
	
	for j := i; j < arrSize; j++ {
		if colArr[j].num > colArr[indx].num {
		indx = j
		}
	}
	maxim = colArr[indx]
	colArr[indx] = colArr[i]
	colArr[i] = maxim
	}
	return colArr
}

}
  
/* --------------------------------------------------------------------------------------------
* Calls sort function with a sortType and prints it out to cmdline
* --------------------------------------------------------------------------------------------*/
func printNums(colArr[10] colNum, arrSize int, sortType int) {
colArr = sort(colArr, arrSize, sortType)
for i := 0; i < arrSize; i++ {
	if colArr[i].num != 0 {
	fmt.Printf("%d %d\n", colArr[i].num, colArr[i].seqLength)
	}
}
}

File: Julia/recursed/colRecur.jl
#!/usr2/local/julia-1.8.2/bin/julia

#!/usr2/local/julia-1.8.2/bin/julia

mutable struct colNum
    num::Int64
    seqLength::Int64
end



function main()

    colArr =  Array{colNum}(undef, 10)
    colArrSize = 1

    lower = parse(UInt64, ARGS[1])
    upper = parse(UInt64, ARGS[2])

    for i in lower:upper

        colValue = colNum(i, 0)
        colValue.seqLength = collatz(colValue.num, colValue.seqLength)
        if (!replaceDuplicate(colArr, colValue, colArrSize))
            if (colArrSize < 11)
                colArr[colArrSize] = colValue
                colArrSize+=1
            else
                location = searchMin(colArr, colArrSize)
                if (colArr[location].seqLength < colValue.seqLength)
                    colArr[location] = colValue
                end
            end
        end
    end
    printNums(colArr, colArrSize, 0)
    printNums(colArr, colArrSize, 1)
end



function collatz(num, count)

    if (num == 1) return count end
    count+=1
    if (num%2 == 0) return collatz(num/2, count) end
    return collatz((num*3)+1, count)

end

function replaceDuplicate(colArr, colValue, arrSize)

    for i in 1:(arrSize-1)
        if (colArr[i].seqLength == colValue.seqLength)
            if (colValue.num <= colArr[i].num) colArr[i] = colValue end
            return true
        end
    end
    return false;
end

function searchMin(colArr, arrSize)

    minSt = colArr[1]
    minSeqLength = 0
    location = 0

    for i in 1:(arrSize-1)
        minSeqLength = min(minSt.seqLength, colArr[i].seqLength)

        if (minSeqLength == colArr[i].seqLength)
            minSt = colArr[i]
            location = i
        end
    end
    return location

end

function sortArr(colArr, arrSize, sortType)

    if (sortType == 0)

        println("Sorted based on sequence length")
        for i in 1:(arrSize-1)
            index = i
        
            for j in i:(arrSize-1)
                if (colArr[j].seqLength > colArr[index].seqLength)
                    index = j
                end
            end
            maxim = colArr[index]
            colArr[index] = colArr[i]
            colArr[i] = maxim
        end

    else
        println("\nSorted based on integer size")
        for i in 1:(arrSize-1)
            index = i
        
            for j in i:(arrSize-1)
                if (colArr[j].num > colArr[index].num)
                    index = j
                end
            end
            maxim = colArr[index]
            colArr[index] = colArr[i]
            colArr[i] = maxim
        end

    end

    return colArr
end


function printNums(colArr, arrSize, sortType)

    colArr = sortArr(colArr, arrSize, sortType)
    for i in 1:(arrSize-1)
        println(colArr[i].num, " ", colArr[i].seqLength)
    end
end

main()